### PR TITLE
Proximity cleanups

### PR DESCRIPTION
**Pull request by peterbarker**
Use singletons in place of passing objects around.

Factor some valid instance code out.

Testing:
 - We test SITL Proximity backend as part of autotest.
 - Morse backend tested.
 - Real RPILidar A2 tested from SITL.

### CODE REVIEW

File: libraries/AP_Proximity/AP_Proximity_RangeFinder.h [hunk 1]
```diff
@@ -31,5 +31,5 @@ class AP_Proximity_RangeFinder : public AP_Proximity_Backend
 
     // upward distance support
     uint32_t _last_upward_update_ms;    // system time of last update distance
-    float _distance_upward;             // upward distance in meters, negative if the last reading was out of range
```
Comment (rmackay9) [inline libraries/AP_Proximity/AP_Proximity_RangeFinder.h]: nitpick: can we align the comments?

Comment (peterbarker) [reply libraries/AP_Proximity/AP_Proximity_RangeFinder.h]: Fixed.

File: libraries/AP_Proximity/AP_Proximity.h [hunk 2]
```diff
@@ -119,7 +119,7 @@ class AP_Proximity
     // The Proximity_State structure is filled in by the backend driver
     struct Proximity_State {
         uint8_t                 instance;   // the instance number of this proximity sensor
-        enum Proximity_Status   status;     // sensor status
+        Status   status;     // sensor status
```
Comment (rmackay9) [inline libraries/AP_Proximity/AP_Proximity.h]: nitpick: misaligned comment here too but let's not worry.

**Comment by rmackay9** [comment]
Merged, thanks!

**Comment by rmackay9** [comment]
this has been added to Copter-4.0.0-rc3 because it was a pre-requisite for this other PR: https://github.com/ArduPilot/ardupilot/pull/12368

**Pull request merged**